Sonic client for Triton server

kpedro88 opened the pull request:
#### PR description:

https://github.com/cms-sw/cmsdist/pull/5868 added triton-inference-server as a CMSSW external. This PR adds the corresponding Sonic client to communicate with Triton servers (following up on #29141 and #29503).

Details:

Triton commands may return an `Error` object. These objects are wrapped and converted into warnings (if the command was not successful). In addition, an unsuccessful command causes the current evaluation to stop (execution will also stop if retries are not allowed).

Optionally, additional reporting from the server side can be enabled using the `verbosity` flag.

The client can select different models on the server by name and version number. The model information (input and output dimensions and sizes) is automatically obtained from the server and can be printed using the `verbosity` flag. Multiple inferences can be done at once by setting the `batchSize` and concatenating all of the input data. (`batchSize` can be changed on a per-event basis if desired.)

A forthcoming PR will add functionality to transmit fixed-point data to a custom FPGA server that recognizes Triton calls. (Currently, models with multiple named inputs and/or outputs and models with variable-size dimensions are not supported. This support will be added in the future if and when relevant examples are available.)

#### PR validation:

Ran the `TritonImageProducer` using the Docker test setup. Tested both success and failure conditions, and observed desired behavior and output.

attn: @violatingcp @drankincms @nhanvtran @jeffkrupa @jmduarte @lgray

kpedro88 committed:
add Sonic client for Triton server

kpedro88 committed:
remove severity mechanism to simplify

cmsbuild commented:
The code-checks are being triggered in jenkins.

kpedro88 commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30318/16302

- This PR adds an extra 20KB to repository

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
A new Pull Request was created by  @kpedro88 (Kevin Pedro) for master.

It involves the following packages:

HeterogeneousCore/SonicCore
HeterogeneousCore/SonicTriton


The following packages do not have a category, yet:

HeterogeneousCore/SonicTriton
Please create a PR for https://github.com/cms-sw/cms-bot/blob/master/categories_map.py to assign category

@makortel, @cmsbuild, @fwyzard can you please review it and eventually sign? Thanks.
@makortel, @rovere this is something you requested to watch as well.
@silviodonato, @dpiparo you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

makortel commented on HeterogeneousCore/SonicTriton/interface/TritonClient.h:
```diff
@@ -0,0 +1,93 @@
+  TritonClient(const edm::ParameterSet& params);
+
+  //helper
+  bool getResults(const std::unique_ptr<nic::InferContext::Result>& result);
```
Why not just
```suggestion
  bool getResults(const nic::InferContext::Result& result);
```
since ownership is not transferred?

kpedro88 replied:
Requires a change elsewhere as well (because of the Triton API), but will do

makortel commented on HeterogeneousCore/SonicTriton/src/TritonClient.cc:
```diff
@@ -0,0 +1,318 @@
+
+template <typename Client>
+TritonClient<Client>::TritonClient(const edm::ParameterSet& params)
+    : Client(),
```
Base class default constructor is not needed (but doesn't harm either).

makortel commented on HeterogeneousCore/SonicTriton/interface/TritonClient.h:
```diff
@@ -0,0 +1,93 @@
+#include "request_grpc.h"
+
+namespace ni = nvidia::inferenceserver;
+namespace nic = ni::client;
```
Quick search suggests that namespace aliases in headers can be a bad idea (because of risks in name collisions, in global namespace in this case).

fwyzard replied:
Agreed...

makortel commented on HeterogeneousCore/SonicTriton/src/TritonClient.cc:
```diff
@@ -0,0 +1,318 @@
+#include <sstream>
+#include <cstring>
+
+namespace ni = nvidia::inferenceserver;
+namespace nic = ni::client;
```
Does
```suggestion
namespace {
  namespace ni = nvidia::inferenceserver;
  namespace nic = ni::client;
}
```
work? If I understood correctly, then the names `ni` and `nic` are not exported outside of this file.

kpedro88 replied:
I did this in the header, but I thought there wasn't a problem with namespace aliases in a .cc file.

makortel replied:
Maybe it isn't a problem after all, the actual symbols use the real namespace name, and whatever shorthands are (should?) not be visible outside of the translation unit.

makortel commented on HeterogeneousCore/SonicTriton/src/TritonClient.cc:
```diff
@@ -0,0 +1,318 @@
+    return status;
+
+  const auto& nicInputs = context_->Inputs();
+  nicInput_ = nicInputs[0];
```
Is `nicInputs` guaranteed to be always non-`empty()`?

kpedro88 replied:
This vector is automatically initialized with the context, based on the model details. I suppose it's possible that the model details on the server could be malformed, so it's worth checking for that case.

makortel commented on HeterogeneousCore/SonicTriton/src/TritonClient.cc:
```diff
@@ -0,0 +1,318 @@
+  auto t1 = std::chrono::high_resolution_clock::now();
+  std::vector<int64_t> input_shape;
+  for (unsigned i0 = 0; i0 < batchSize_; i0++) {
+    float* arr = &(this->input_.data()[i0 * nInput_]);
```
Why go via `vector::data()`? Wouldn't `operator[]()` suffice?
```suggestion
    float* arr = &(this->input_[i0 * nInput_]);
```

makortel commented on HeterogeneousCore/SonicTriton/src/TritonClient.cc:
```diff
@@ -0,0 +1,318 @@
+    if (!status)
+      return status;
+    const float* lVal = reinterpret_cast<const float*>(r0);
+    std::memcpy(&this->output_[i0 * nOutput_], lVal, nOutput_ * sizeof(float));
```
Given that the copy goes via `memcpy()`, the `reinterpret_cast` above is unnecessary, no? (`std::copy()` would be different)

cmsbuild commented:
-1

Tested at: ca83125faa8cbbb3145d6e000dd3e0adf126c007

CMSSW: CMSSW_11_2_X_2020-06-19-1100
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-3b612c/7215/summary.html

I found follow errors while testing this PR

**Failed tests:  Build  ClangBuild**

* **Build**:

I found compilation warning when building:  See details on the summary page.

* **Clang**:

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.

cmsbuild commented:
Comparison not run due to Build errors (RelVals and Igprof tests were also skipped)

makortel commented on HeterogeneousCore/SonicTriton/interface/TritonClient.h:
```diff
@@ -0,0 +1,93 @@
+  std::shared_ptr<nic::InferContext::Input> nicInput_;
+};
+
+typedef TritonClient<SonicClientSync<std::vector<float>>> TritonClientSync;
+typedef TritonClient<SonicClientPseudoAsync<std::vector<float>>> TritonClientPseudoAsync;
+typedef TritonClient<SonicClientAsync<std::vector<float>>> TritonClientAsync;
```
How about
```suggestion
using TritonClientSync = TritonClient<SonicClientSync<std::vector<float>>>;
using TritonClientPseudoAsync = TritonClient<SonicClientPseudoAsync<std::vector<float>>>;
using TritonClientAsync = TritonClient<SonicClientAsync<std::vector<float>>>;
```
?

kpedro88 replied:
Not opposed, but what is the particular benefit?

makortel replied:
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rt-using

makortel replied:
> **Reason** Improved readability: With `using`, the new name comes first rather than being embedded somewhere in a declaration. Generality: using can be used for template aliases, whereas typedefs can’t easily be templates. Uniformity: `using` is syntactically similar to `auto`.

fwyzard replied:
IIUC the c++ standard, the `typedef` or `using` statements do not count as explicit template instantiation declarations.

So there should also be the lines
```c++
// explicit template instantiation declarations
extern template class TritonClient<SonicClientSync<std::vector<float>>>;
extern template class TritonClient<SonicClientAsync<std::vector<float>>>;
extern template class TritonClient<SonicClientPseudoAsync<std::vector<float>>>;
```
to match the definitions in the .cc file.

kpedro88 replied:
I was under the impression that `extern template` was intended to *prevent* implicit instantiation, because there is an explicit instantiation elswhere: https://isocpp.org/wiki/faq/cpp11-language-templates. Since the class template introduced here is not fully implemented, there should not be any chance of implicit instantiation. But perhaps my understanding is incomplete.

fwyzard replied:
But you do have an explicit template instantiation elsewhere, at the bottom of the `HeterogeneousCore/SonicTriton/src/TritonClient.cc` file.

What am I missing ?

kpedro88 replied:
My point is that *preventing* implicit instantiation in the header file is not necessary, because the class cannot be compiled using just the information in the header file. `extern template` only affects compilation time, as far as I understand.

I tried compiling a few times with and without these lines, and I don't noticed a significant difference in the compilation time.

(Also, for future reference, if these lines are added, `template <> void TritonClientAsync::evaluate();` is needed before them to avoid an "explicit specialization after instantiation" error.)

makortel replied:
> My point is that _preventing_ implicit instantiation in the header file is not necessary, because the class cannot be compiled using just the information in the header file.

The template class can be instantiated, and calling inlined functions works fine. Calling out-of-line functions also compiles, but leads to linker errors if the out-of-line function definition (for the template argument) is not visible to the linker.

kpedro88 replied:
Is this comment in support of using `extern template class`? (I only find one other instance of it in all of CMSSW.)

makortel replied:
"maybe"? I don't have clear preference in either way, just wanted to point out that a class template with just function declarations can be instantiated, and everything seems to work "fine" until a declared-only function is called.

kpedro88 replied:
I was curious, so I ran the compilation from scratch 100 times with and without this change on a dedicated machine. It decreases the compile time for this package by 4% on average, with a significance of 4.3 sigma. So it's definitely a micro-optimization in this case, but it does have an impact. I'll make the requested change.

makortel commented:
Do I guess correctly that crafting tests for these is a bit challenging (would need a Triton server to talk to)?

fwyzard commented on HeterogeneousCore/SonicTriton/src/TritonClient.cc:
```diff
@@ -0,0 +1,318 @@
+      return status;
+  }
+  auto t2 = std::chrono::high_resolution_clock::now();
+  if (!this->debugName_.empty())
```
Here and below: why the need to explicitly write `this->` ?

kpedro88 replied:
`TritonClient` is a class template that inherits from its template argument: https://github.com/cms-sw/cmssw/blob/ca83125faa8cbbb3145d6e000dd3e0adf126c007/HeterogeneousCore/SonicTriton/interface/TritonClient.h#L19-L20

The compiler doesn't check the base class for nondependent fields (https://isocpp.org/wiki/faq/templates#nondependent-name-lookup-members). Using `this->` makes it dependent, so the compiler can find inherited members and methods.

I did find a few cases where `this->` is not needed, and I'll remove those.

(There's a separate case in `TritonClientAsync::evaluate()` where `this` is passed to a lambda, and therefore `this->` must be used inside the lambda.)

fwyzard replied:
I see, thanks for the explanation.

(I think if you explicitly capture `this` by value in a lambda, you can reference methods and data members directly, without `this->`, unless they are shadowed by some other variable.)

kpedro88 replied:
Thanks, I've removed all the `this->`s in the lambda.

fwyzard commented on HeterogeneousCore/SonicTriton/interface/TritonClient.h:
```diff
@@ -0,0 +1,93 @@
+    descClient.add<unsigned>("batchSize");
+    descClient.add<std::string>("address");
+    descClient.add<unsigned>("port");
+    descClient.add<unsigned>("timeout");
```
I think some of the parameters (e.g. `address`, `port`, `timout`, ...) should be `untracked`, since they are not supposed to impact the physics results

kpedro88 replied:
I'm willing to make this change, but it's not impossible that the choice of server could impact the physics results. For example, floating point math might give slightly different results.

fwyzard replied:
Honestly, that would just be a limitation of the implementation - like we can get slightly different rersults due to different CPU vendors.

But if this is ever used in production, I don't think one can assume that a single server should be able to handle the requests for a whole dataset: for example, parts of a dataset could be processed in different sites, which then would likely need to use different servers.

kpedro88 replied:
Based on this idea, I think the only tracked parameters should be: nInput, nOutput, modelName, modelVersion. I will change the rest to untracked.

makortel replied:
(thinking out loud) On the other hand, we are tracking in the provenance the choices of `SwitchProducer`, and we are going to add tracking of the chosen "vector architecture" (the recent scram feature). Further on the other hand, we do not track the exact CPU type (e.g. Intel vs AMD for non-standard fast math) nor `$SCRAM_ARCH` (e.g. amd64 vs arm vs ppc), and one could easily argue the exact server to be similar to either of those.

Anyway, keeping them untracked is fine for me now.

kpedro88 committed:
first round of comments

kpedro88 committed:
remove some unneeded this->

kpedro88 committed:
server params untracked

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30318/16334

- This PR adds an extra 20KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30318/16334/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30318/16334/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

kpedro88 committed:
extern template, code format

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30318/16338

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #30318 was updated. @makortel, @cmsbuild, @fwyzard can you please check and sign again.

kpedro88 commented:
@makortel indeed, a standardized testing environment is difficult to establish at the moment. I used an FNAL GPU server for my private tests of this PR, but this is a development server, so it can't be dedicated to running tests for CMSSW in general.

Testing servers for SONIC should be considered as a longer-term item, going toward the goal of production readiness.

As an interim option, it is possible to launch a local server using a docker image provided by Nvidia. This server can run inference on the CPU if a GPU is not available. Together with an appropriate model file, this could be used for an "on-the-fly" testing environment. However, it's not clear to me if this is too heavy to be a unit test in CMSSW.

makortel commented:
> As an interim option, it is possible to launch a local server using a docker image provided by Nvidia. This server can run inference on the CPU if a GPU is not available. Together with an appropriate model file, this could be used for an "on-the-fly" testing environment. However, it's not clear to me if this is too heavy to be a unit test in CMSSW.

Maybe @smuzaffar can comment?

Either way, I'm fine if the tests are addressed in the future.

smuzaffar commented:
cmssw tests already run under docker/singularity. I am not sure if one can launch docker within docker.

kpedro88 committed:
add docker-based test setup

kpedro88 committed:
get contexts once, when constructing client

kpedro88 committed:
move more into constructor

kpedro88 committed:
get and use model info from server, check for unsupported cases

kpedro88 committed:
code format

cmsbuild commented:
The code-checks are being triggered in jenkins.

kpedro88 commented:
Info on the latest updates:

* I added standalone test instructions based on https://docs.nvidia.com/deeplearning/triton-inference-server/master-user-guide/docs/run.html. I do not think this can be included as a CMSSW unit test: the model data are 123MB, and the Nvidia Triton server Docker image is 6GB. (Docker within Docker is technically possible, but might be tricky.) However, anyone can set this up on their own machine.
* In the process of setting up the standalone tests, I realized that some configuration parameters could be obtained directly from the model info on the server (better enforcing consistency). This is now implemented, and the relevant model info can also be printed for the user.
* I found and fixed some inefficiencies (server contexts obtained for every call, when they can just be obtained once and reused). This reduced unnecessary overhead (the difference between "Remote time" and "Client time" in the test example dropped by ~3x).
* I also realized there are some model features that we aren't currently supporting: multiple named inputs/outputs, and variable-size dimensions. The client now tests for these features and throws an exception if it finds them. If and when we have examples that use these features (we currently don't), we can generalize the client to support them (might be non-trivial in some cases).

@makortel @fwyzard your feedback is welcome.

kpedro88 commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30318/16371

- This PR adds an extra 32KB to repository

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #30318 was updated. @makortel, @cmsbuild, @fwyzard can you please check and sign again.

kpedro88 committed:
fix typos

cmsbuild commented:
The code-checks are being triggered in jenkins.

kpedro88 commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30318/16372

- This PR adds an extra 44KB to repository

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #30318 was updated. @makortel, @cmsbuild, @fwyzard can you please check and sign again.

cmsbuild commented:
+1
Tested at: fa33239c26c7c0c3072a0c1b5d60916d8d17e57a
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-3b612c/7301/summary.html
CMSSW: CMSSW_11_2_X_2020-06-23-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-3b612c/7301/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 36
 - DQMHistoTests: Total histograms compared: 2778811
 - DQMHistoTests: Total failures: 4
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2778757
 - DQMHistoTests: Total skipped: 50
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 35 files compared)
 - Checked 152 log files, 16 edm output root files, 36 DQM output files

makortel commented on HeterogeneousCore/SonicTriton/test/imageTest_cfg.py:
```diff
@@ -0,0 +1,75 @@
+    )
+
+if options.threads>0:
+    if not hasattr(process,"options"): process.options = cms.untracked.PSet()
+    process.options.numberOfThreads = cms.untracked.uint32(options.threads)
+    process.options.numberOfStreams = cms.untracked.uint32(options.streams if options.streams>0 else 0)
```
Nowadays `process.options` always exists with the supported parameters, so this can be just
```suggestion
    process.options.numberOfThreads = options.threads
    process.options.numberOfStreams = options.streams
```
(assuming no-one would set `streams < 0`)

makortel reviewed (commented):
Thanks Kevin for the testing recipe (even if I didn't try it). I hope we could find a solution to exercise it some time in the future.

kpedro88 committed:
address more comments

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30318/16489

- This PR adds an extra 52KB to repository

cmsbuild commented:
Pull request #30318 was updated. @makortel, @cmsbuild, @fwyzard can you please check and sign again.

kpedro88 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

kpedro88 commented:
see https://github.com/cms-sw/cms-bot/pull/1332 for category

cmsbuild commented:
+1
Tested at: da50dccf8a30ffeae5e87faf4eff79c80907e0b1
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-3b612c/7351/summary.html
CMSSW: CMSSW_11_2_X_2020-06-25-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-3b612c/7351/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 36
 - DQMHistoTests: Total histograms compared: 2778915
 - DQMHistoTests: Total failures: 4
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2778861
 - DQMHistoTests: Total skipped: 50
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 35 files compared)
 - Checked 152 log files, 16 edm output root files, 36 DQM output files

fwyzard commented:
Do we really need to add a new package for a single class, a single source file and its corresponding header file ?
What is the plan to add more clients ?
If the others follow a similar approach, wouldn't it make more sense to add a single `HeterogeneousCore/SonicClients` package ?

kpedro88 commented:
The reason I've made this a separate package is to split up the dependencies. This client depends on the `triton-inference-server` external, while clients for other servers will have their own, separate dependencies. Would you prefer a different approach?

fwyzard commented:
I see your point - that's why I'm wondering if you already know what
further packages may look like ?

 If you foresee only 2-3 more, and the others may possibly contain more
code, then the approach of one package per backend should be fine.

 If you foresee having something like 10 packages, each with a single
class, maybe we can think of a more compact approach.

kpedro88 commented:
Here's what I currently foresee:
* TFServing-based client (code already exists, I just need to have the energy to struggle with bazel to make the external acceptable for cmsdist)
* Apache IPC (on the wish list)
* Something to support REST (early private tests were discouraging from a speed point of view, so this is not currently a priority)

I want to leave open the possibility to add other server technologies if they become interesting. Though, to me, Triton seems like the clear winner at the moment (but the use cases for Apache IPC are different).

The SonicTriton package could also potentially host the associated producers in the plugins directory, though I don't know what reco thinks about this organization. @slava77 can you comment?

slava77 commented:
> The SonicTriton package could also potentially host the associated producers in the plugins directory, though I don't know what reco thinks about this organization. @slava77 can you comment?

there's nothing reco-specific about plugins having to be in /plugins directory; it's probably core or general CMSSW guidance.

kpedro88 commented:
I was more wondering about having plugins kept in the SonicTriton package that run in reco.

makortel commented:
I had thought so far that the reco plugins would go in appropriate packages outside `HeterogeneousCore` (and that is the current approach with CUDA). On the other hand, there is a clear tension between trying to avoid loading unnecessary dependency libraries, and proliferation of packages or plugin libraries.

For CUDA we sort-of compromised allowing CUDA runtime to be depended on always, because it is simpler and does not cause problems (now, remains to be seen if this decision comes back to bite us). Following that approach one package for Sonic with all the backends could be a way to go. But we should be fairly certain that loading all the dependencies all the time would not cause any ill effects like (symbol) conflicts, significant memory use, extra non-sleeping threads.

@Dr15Jones would you have any thoughts on this?

fwyzard commented:
Thanks for the plan, Kevin.

With 3-4 back-ends currently foreseen, I don't see it as a problem to have
a package each.

kpedro88 commented:
@makortel for the immediate future, we probably want to keep triton-inference-server dependencies isolated. There are some minor thread-safety issues in the version we're currently using. I'm going to work directly with the developers to address this. (They're also in the process of moving to a new/updated API; once that is stable, our code will be updated accordingly.)

But the client library is overall not very large, so in the future it may not be too much overhead if the dependency spreads out.

makortel commented:
This PR looks good now, I don't have any further comments. @fwyzard do you?

fwyzard commented:
I don't think so - I'll have an other look and sign later today.

kpedro88 commented:
ping @fwyzard, please let me know if I should take care of anything else

fwyzard commented on HeterogeneousCore/SonicTriton/test/fetch_model.sh:
```diff
@@ -0,0 +1,17 @@
+mkdir -p $MODEL_DIR
+cd $MODEL_DIR
+
+wget https://github.com/NVIDIA/triton-inference-server/raw/v${TRITON_VERSION}/docs/examples/model_repository/resnet50_netdef/config.pbtxt
```
Just a suggestion, feel free to ignore it: `wget` is not bundled with CMSSW, while `curl` is, so you may want to use the latter in the test to make sure it's always available.

kpedro88 replied:
It's a reasonable suggestion; I'll consider this for a followup PR.

fwyzard commented on HeterogeneousCore/SonicTriton/src/TritonClient.cc:
```diff
@@ -0,0 +1,405 @@
+  //non-blocking call
+  auto t1 = std::chrono::high_resolution_clock::now();
+  status =
+      wrap(context_->AsyncRun([t1, start_status, this](nic::InferContext* ctx,
```
out of curiosity: why do you prefer to measure `t1` outside the lambda and then capture it, instead of directly measuring it inside the lambda ?

the latter would also make it easier to measure it only if `(!debugName_.empty())`, since it is only used in that case.

kpedro88 replied:
`AsyncRun()` performs some operations before executing the callback, and we would like to include those in our profiling.

fwyzard commented:
+heterogeneous

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo (and backports should be raised in the release meeting by the corresponding L2)

fwyzard commented:
(we should have chosen a category name easier to spell...)

kpedro88 commented:
@silviodonato it would be great to get this into pre2

silviodonato commented:
+1
